src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,250 @@
+import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import java.sql.Types
+import java.sql.PreparedStatement
+import java.sql.ResultSet
+import java.time.Instant
+import java.time.LocalDate
+import java.time.OffsetDateTime
+import java.time.ZoneId
+import java.util.*
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantLock
+import java.util.function.Function
+import java.util.function.Supplier
+
+/**
+ *
+ */
+
+const val MAX_DURATION_MILLIS = 60_000L
+const val DATA_DELETION_RATE = 30_000L
+
+class BackgroundExpiredDataDeletionService(
+        hazelcastInstance: HazelcastInstance,
+        private val indexerConfiguration: IndexerConfiguration,
+        private val hds: HikariDataSource,
+        private val elasticsearchApi: ConductorElasticsearchApi
+) {
+    companion object {
+        private val logger = LoggerFactory.getLogger(BackgroundExpiredDataDeletionService::class.java)!!
+    }
+
+    private val propertyTypes: IMap<UUID, PropertyType> = hazelcastInstance.getMap(HazelcastMap.PROPERTY_TYPES.name)
+    private val entitySets: IMap<UUID, EntitySet> = hazelcastInstance.getMap(HazelcastMap.ENTITY_SETS.name)
+    private val expirationLocks: IMap<UUID, Long> = hazelcastInstance.getMap(HazelcastMap.EXPIRATION_LOCKS.name)
+
+    //necessary??????
+    init {
+        expirationLocks.addIndex(QueryConstants.THIS_ATTRIBUTE_NAME.value(), true)
+    }
+
+    private val taskLock = ReentrantLock()
+
+    @Suppress("UNCHECKED_CAST", "UNUSED")
+    @Scheduled(fixedRate = MAX_DURATION_MILLIS)
+    fun scavengeIndexingLocks() {
+        expirationLocks.removeAll(
+                Predicates.lessThan(
+                        QueryConstants.THIS_ATTRIBUTE_NAME.value(),
+                        System.currentTimeMillis()
+                ) as Predicate<UUID, Long>
+        )
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = DATA_DELETION_RATE)
+    fun deleteExpiredDataFromEntitySets() {
+        logger.info("Starting background expired data deletion task.")
+        //Keep number of expired data deletion jobs under control
+        if (taskLock.tryLock()) {
+            try {
+                if (indexerConfiguration.backgroundExpiredDataDeletionEnabled) {
+                    val w = Stopwatch.createStarted()
+                    //We shuffle entity sets to make sure we have a chance to work share and index everything
+                    //lock entity sets we are working on
+                    val lockedEntitySets = entitySets.values
+                            .shuffled()
+                            .filter { tryLockEntitySet(it) } //filters out entitysets that are already locked, and the entitysets we're working on are now locked in the IMap
+                            .filter { it.name != "OpenLattice Audit Entity Set" } //TODO: Clean out audit entity set from prod
```

Comment:
I don't think this filter is necessary (I think this had to do with some old startup/EDM issue and is probably/hopefully not necessary for indexing either anymore)

Reply:
removed